java-checks/src/main/java/org/sonar/java/checks/FieldModifierCheck.java

```diff
@@ -31,10 +32,12 @@
 import org.sonar.plugins.java.api.tree.VariableTree;
 
 import java.util.List;
+import java.util.function.Predicate;
 
 @Rule(key = "S2039")
 public class FieldModifierCheck extends IssuableSubscriptionVisitor {
 
+  private static final Predicate<ModifiersTree> MODIFIERS_TREE_PREDICATE = new VisibleForTestingPredicate().negate();
```

Comment:
OK, let's put aside the unit/integration part of the discussion.

After thinking a bit more about your approach, I realized that there is a simpler way to check presence of the annotation, by relying 100% on the semantic API (and fully unit tested! OK, I stop.)

In the `isNotVisibleForTesting` method, instead of using your predicate to test the presence of the annotation, you should do the exact same thing using the symbol metadata. I'll add a comment there.